cilium: bandwidth manager e2e tests

borkmann opened the pull request:
See commit description.

borkmann committed:
cilium, test: add various netperf e2e tests

Add {Pod,Host} to Pod netperf tests over VMs with different bandwidth
limits in direct routing and tunneling mode.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on test/k8sT/manifests/demo_bw.yaml:
```diff
@@ -0,0 +1,120 @@
+    spec:
+      containers:
+      - name: netperf-50
+        image: cilium/netperf:latest
```
Could you tag the image and and add it to `cilium/packer-ci-build/provision/pull-images.sh`?

borkmann replied:
Added PR here https://github.com/cilium/packer-ci-build/pull/234

brb commented on test/k8sT/Bandwidth.go:
```diff
@@ -0,0 +1,146 @@
+			pods, err := kubectl.GetPodNames(helpers.DefaultNamespace, clientPodLabel)
+			ExpectWithOffset(1, err).Should(BeNil(), "cannot retrieve pod names by filter %q",
+				clientPodLabel)
+			cmd := helpers.NetperfTerse(netperfWrapperBase, targetIP, helpers.TCP_MAERTS, "-l 30 -I 99,5")
```
Did you add the script to the cilium/netperf:latest image? If not, then it won't be available, and thus the exec at the line 117 will fail.

borkmann committed:
bpf: call edt_sched_departure from overlay egress

Paul reported that rate-limiting with tunneling enabled currently does
not work. This is due to queue_mapping being trashed upon dev_queue_xmit()
on the vxlan/geneve dev. Instead of mapping this to some other skb meta
data in the hope it doesn't get overridden during vxlan/geneve xmit, just
set the timestamp from there. Given they have noqueue it gets close enough;
the reported netperf rates are close enough.

Reported-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

christarazi commented on test/k8sT/Bandwidth.go:
```diff
@@ -0,0 +1,170 @@
+					ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
+						"Request from %s pod to pod with IP %s failed", pod, targetIP)
+					if rate > 0 {
+						Expect(res.InRange(rate, maxRateDeviation)).To(BeNil(), "Rate mismatch")
```
This should be `ExpectWithOffset` to be consistent with the rest of the expectations in this function.

christarazi commented on test/k8sT/Bandwidth.go:
```diff
@@ -0,0 +1,170 @@
+				podIPs, err := kubectl.GetPodsIPs(helpers.DefaultNamespace, label)
+				ExpectWithOffset(1, err).Should(BeNil(), "Cannot retrieve pod IPs for %s",
+					label)
+				Expect(len(podIPs)).To(Equal(int(1)), "Expected pod IPs mismatch")
```
Should be `ExpectWithOffset` here too

christarazi reviewed (approved):
LGTM, minor changes required, but don't need a second review from me.

borkmann committed:
bpf: reset edt aggregate for nodeport traffic that is remotely handled

Reset it to zero so that rx queue mapping won't accidentally select a
known aggregate given we don't rate-limit these. This is for fwd direction
where backend is remote as well as reverse direction in case of SNAT where
reply hits the middle node again.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
